src/braket/experimental_capabilities/experimental_capability_context.py

```diff
@@ -24,40 +34,110 @@ class ExperimentalCapabilityContextError(Exception):
+        )
+
+    @_is_enabled.register
+    def _(self, experimental_capabilities: None):
+        return self._all_enabled
+
+    @_is_enabled.register
+    def _(self, experimental_capabilities: ExperimentalCapability):
+        return (
+            self._all_enabled or experimental_capabilities in self._currently_enabled_capabilities
+        )
 
-    def disable(self) -> None:
-        """Disable all experimental capabilities."""
-        self._is_enabled = False
+    @_is_enabled.register
+    def _(self, experimental_capabilities: Iterable):
+        return self._all_enabled or set(experimental_capabilities).issubset(
+            self._currently_enabled_capabilities
+        )
 
-    def set_state(self, state: bool) -> None:
+    def enable(self, experimental_capabilities: set[ExperimentalCapability] | None = None) -> None:
+        """Enable all experimental capabilities."""
+        if experimental_capabilities is None:
+            self._all_enabled = True
+        else:
+            capabilities_set = set(experimental_capabilities)
+            if ExperimentalCapability.ALL in capabilities_set:
+                if len(capabilities_set) > 1:
+                    warnings.warn(
+                        '"ALL" was passed in along with other explicit capabilities. '
+                        "All experimental capabilities will be enabled. If this is a "
+                        "mistake, please check your usage of the experimental capabilities "
+                        "context manager.",
+                        stacklevel=1,
+                    )
+                self._all_enabled = True
+            else:
+                self._currently_enabled_capabilities.update(capabilities_set)
+
+    def disable(self, experimental_capabilities: set[ExperimentalCapability] | None = None) -> None:
+        """Disable all specified experimental capabilities; defaults to disabling all.
+
+        Args:
+            experimental_capabilities: the experimental capabilities to disable.
+        """
+        if experimental_capabilities is None:
+            self._all_enabled = False
+            self._currently_enabled_capabilities.clear()
+        else:
+            if ExperimentalCapability.ALL in experimental_capabilities:
+                self._all_enabled = False
+            self._currently_enabled_capabilities.difference_update(experimental_capabilities)
+
+    def set_state(
+        self, enabled_capabilities: Iterable[ExperimentalCapability] | None = None
+    ) -> None:
         """Set the state of all experimental capabilities.
 
         Args:
-            state: The state to set. True to enable all capabilities, False to disable all.
+            enabled_capabilities: the state of the experimental capability context in terms of the
+            experimental capabilities to enable. If None, default to ["ALL"]
         """
-        self._is_enabled = bool(state)
+        self.enable(enabled_capabilities)
```

Comment:
This is not setting the state. If the `_currently_enabled_capabilities={X}` and we `set_state([Y])`, `_currently_enabled_capabilities` becomes `{X, Y}` while we expect `{Y}`. 

This becomes important in nested context where 
```
with EnableExperimentalCapability([A,B]):
    # {A, B} enabled
    with EnableExperimentalCapability([C]):
           # {A, B, C} enabled
    # {A, B} enabled
```

Reply:
Also, what if the previous state is 
```
_all_enabled = True
_currently_enabled_capabilities = {}
```
It won't set `_all_enabled`.

Reply:
Now I see your implementation in `__exit__` below. I feel like we should either (1) fully implement `set_state()` or (2) replace the implementation in `__exit__` with
```
GLOBAL_EXPERIMENTAL_CAPABILITY_CONTEXT.disable(self._current_enabled_capabilities)
GLOBAL_EXPERIMENTAL_CAPABILITY_CONTEXT.enable(self._previous_enabled_capabilities)
```
and remove `def set_state()`.